src/main/java/com/iota/iri/LedgerValidator.java

```diff
@@ -198,103 +187,89 @@ private void updateConsistentHashes(final Snapshot snapshot, Hash tip, int index
-                }
+        StateDiffViewModel stateDiffViewModel;
+        milestone.latestSnapshot.rwlock.writeLock().lock();
+        MilestoneViewModel snapshotMilestone = MilestoneViewModel.firstWithSnapshot(tangle);
+        while (snapshotMilestone != null) {
+            stateDiffViewModel = StateDiffViewModel.load(tangle, snapshotMilestone.getHash());
+            if(Snapshot.isConsistent(milestone.latestSnapshot.patch(stateDiffViewModel.getDiff()))) {
+                milestone.latestSnapshot.merge(stateDiffViewModel.getDiff(), snapshotMilestone.index());
+                consistentMilestone = snapshotMilestone;
+                snapshotMilestone = snapshotMilestone.nextWithSnapshot(tangle);
             }
         }
+        milestone.latestSnapshot.rwlock.writeLock().unlock();
         return consistentMilestone;
     }
 
-    public boolean updateSnapshot(Snapshot latestSnapshot, MilestoneViewModel milestone) throws Exception {
-        TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle, milestone.getHash());
-        synchronized (latestSnapshot.snapshotSyncObject) {
-            final int lastSnapshotIndex = latestSnapshot.index();
-            final int transactionSnapshotIndex = transactionViewModel.snapshotIndex();
-            boolean hasSnapshot = transactionSnapshotIndex != 0;
-            if(!hasSnapshot) {
-                Hash tail = transactionViewModel.getHash();
-                Map<Hash, Long> currentState = getLatestDiff(latestSnapshot, tail, lastSnapshotIndex, true);
-                hasSnapshot = currentState != null && latestSnapshot.patch(currentState, milestone.index()).isConsistent();
-                if (hasSnapshot) {
-                    updateSnapshotMilestone(milestone);
-                    synchronized (latestSnapshot.approvalsSyncObject) {
-                        latestSnapshot.approvedHashes.clear();
-                    }
-                    StateDiffViewModel stateDiffViewModel;
-                    stateDiffViewModel = new StateDiffViewModel(currentState, milestone.getHash());
-                    if(currentState.size() != 0) {
-                        stateDiffViewModel.store(tangle);
-                    }
-                    latestSnapshot.merge(latestSnapshot.patch(stateDiffViewModel.getDiff(), milestone.index()));
+    public boolean updateSnapshot(Snapshot latestSnapshot, MilestoneViewModel milestoneVM) throws Exception {
+        TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle, milestoneVM.getHash());
+        this.milestone.latestSnapshot.rwlock.writeLock().lock();
+        final int transactionSnapshotIndex = transactionViewModel.snapshotIndex();
+        boolean hasSnapshot = transactionSnapshotIndex != 0;
+        if(!hasSnapshot) {
+            Hash tail = transactionViewModel.getHash();
+            Set<Hash> visitedHashes = new HashSet<>();
```

Comment:
`visitedHashes` is never used in the function.
better just pass straight to `getLatestDiff`. otherwise, it's confusing: has a name -> has a purpose.